### PR TITLE
Usability improvement: declaring exclusive content of repositories

### LINKED ISSUE #9515: Allow repository filtering API to express exclusivity
### Context

Today, if you want to express that a dependency _must_ be found only in one repository and not others, you have to:

- declare an `include` for this dependency for the repository it should use
- declare `exclude` for this dependency on all repositories it should not use

This can be cumbersome if there are lots of dependencies.

Some ideas:

#### Suggestion by @ldaley

```
content {
   exclusivelyIncludeGroup("org.foo") // means do not try to look other repositories for this group
}
```

#### Alternative name

```
content {
   singleSourceForGroup("org.foo") // means do not try to look other repositories for this group
}
```

#### Grouped exclusivity

```
content {
    exclusiveForOtherRepositories {
        includeGroup ...
    }
}
```

Note that this can become a bit complicated if you have conflicts, in particular between regex include/excludes and repositories. Then it becomes order dependent...

We cannot rely on declarations at the dependency usage site, because repositories are used at the _consumer_ side (both for reproducibility and security). Meaning that we _cannot_ have:

```
dependencies {
     implementation("...") from jcenter()
}
```

Because this creates a whole category of new problems (what if a transitive says nothing? What if a transitive has a different version? What if a transitive says a different repo? What if a repo is spoofed?)

### PR DESCRIPTION
**Pull request by melix**
### Context

Fixes #9515

See commits for details.

Example syntax:

```groovy
repositories {
    // This repository will _not_ be searched for artifacts in my.company
    // despite being declared first
    jcenter()
    exclusiveContent {
        forRepository {
            maven {
                url "https://repo.mycompany.com/maven2"
            }
        }
        filter {
            // this repository *only* contains artifacts with group "my.company"
            includeGroup "my.company"
        }
    }
}
```

### CODE REVIEW

**Comment by vlsi** [comment]
Just wondering, can the DSL be like the following?

```gradle
repositories {
    // This repository will _not_ be searched for artifacts in my.company
    // despite being declared first
    jcenter()
    maven {
        url "https://repo.mycompany.com/maven2"
        exclusiveContent {
            // this repository *only* contains artifacts with group "my.company"
            includeGroup "my.company"
        }
    }
}
```

**Comment by melix** [comment]
> Just wondering, can the DSL be like the following?

In theory, yes, it could, but I think it wouldn't be good for 2 reasons:

1. you're saying something about _multiple_ repositories, not just one, so putting it _within_ the repository declaration is misleading
2. it would force keeping state about all repositories in every repository instance (each repository would need to keep track of the other repositories created with the same container), which is not great.

**Comment by vlsi** [comment]
> you're saying something about multiple repositories, not just one, so putting it within the repository declaration is misleading

That depends.

I've been smoke-testing my Gradle plugin, and now I have the following:

```kotlin
    repositories {
        // Here's a repository with staged plugin:
        maven("https://oss.sonatype.org/content/repositories/comgithubautostyle-1012")
        // All the rest plugins
        gradlePluginPortal()
    }
```

I was trying to specify like "do not try resolving com.github.autostyle from plugin portal", and I tried to use IDEA's autocomplete for that.
The problem was Gradle was trying to fetch each plugin from `comgithubautostyle-1012` which took time.

Unfortunately, the exclusive content feature is not there, so I ended up with just reordering the repositories:

```kotlin
    repositories {
        gradlePluginPortal()
        maven("https://oss.sonatype.org/content/repositories/comgithubautostyle-1012")
    }
```

Of course, it is not a big deal for me, however, as a end-user, I expect that I could specify "well, this group/artifact/version can only be present in this repository, don't even try the others".

It is kind of sad that your suggested `exclusiveContent` is not a trivial flip when one wants to add an exclusive content filter to an already present repository in code.


Well, it might be alleviated by adding IDEA's intentions to flip between the two representations.
However, it is sad that "adding exclusive filter" does not look like just adding `exclusiveContent {...}`

---

Note: what if filters overlap?
AFAIK it would be ok (useful and understandable) to have the following:

```kotlin
repositories {
    // This repository will _not_ be searched for artifacts in my.company
    // despite being declared first
    jcenter()
    maven {
        url "https://repo.mycompany.com/maven2/releases"
        exclusiveContent {
            includeGroup "my.company"
        }
    }
    maven {
        url "https://repo.mycompany.com/maven2/tools"
        exclusiveContent {
            includeGroup "my.company"
        }
    }
}
```

The logic as follows: if an artifact matches at least one "exclusive" filter, then it must be resolved from the matching repositories in the order of their declaration.

In other words, `my.company:tool` is resolved from `releases` repo first, then from `tools`.
It does sound composable with the existing `content { .. }` feature, and it is discoverable with auto-complete inside `maven {...}` block.

**Comment by vlsi** [comment]
>it would force keeping state about all repositories in every repository instance (each repository would need to keep track of the other repositories created with the same container), which is not great.

Do you really think so?

I guess the implementation could be like:
1) If repository is using non-trivial exclusive content filter, then it is added to the list of `repositories with exclusive content activated`.
2) At resolution time, Gradle could check if actual GAV matches one of the `exclusive content filters`
3) If there's a match, then Gradle uses just matching exclusive repositories. Otherwise it uses regular resolution.

I do not see why cross-references between repositories are needed.

**Comment by melix** [comment]
> If a repository is using a non-trivial exclusive content filter, then it is added to the list of repositories with exclusive content activated.

To be added to such a list it would need to be passed the list _or_ its container. But again, I don't think it's a good idea to put at the _repository level_ something that says something about _all repositories_.

**Comment by vlsi** [comment]
> repository level something that says something about all repositories.

What is the solution for the following then?

```
repositories {
    // This repository will _not_ be searched for artifacts in my.company
    // despite being declared first
    jcenter()
    mavenCentral()
    google()
    maven {
        url "https://repo.mycompany.com/maven2/releases"
        exclusiveContent {
            includeGroup "my.company"
        }
    }
    maven {
        url "https://repo.mycompany.com/maven2/tools"
        exclusiveContent {
            includeGroup "my.company"
        }
    }
}
```

**Comment by melix** [comment]
> It is kind of sad that your suggested exclusiveContent is not a trivial flip when one wants to add an exclusive content filter to an already present repository in code.

I don't understand what you mean here. The repository in `forRepository` doesn't _have_ to be created in the `forRepository` block:

```
repositories {
   val myRepo = maven { ... }
   exclusiveContent {
      forRepository { myRepo }
      filter { ... }
   }
}
```

**Comment by melix** [comment]
```
repositories {
    // This repository will _not_ be searched for artifacts in my.company
    // despite being declared first
    jcenter()
    mavenCentral()
    google()
    maven {
        url "https://repo.mycompany.com/maven2/releases"
        exclusiveContent {
            includeGroup "my.company"
        }
    }
    maven {
        url "https://repo.mycompany.com/maven2/tools"
        exclusiveContent {
            includeGroup "my.company"
        }
    }
}
```

I would argue that in this case the content is _not_ exclusive, because precisely it can be found in multiple repositories. One problem today with composing exclusivity like this is that the overlapping of filters. For regular expressions typically you can't tell if it's going to match another (well you can but then I need to switch jobs for a few weeks to rewrite an automaton library).

On the other hand it should be pretty straightforward to implement it with the version I suggested, by allowing more than one repository:

```
repositories {
    exclusiveContent {
        forRepository { repo1 }
        forRepository { repo2 }
    }
}
```

I think it would be a good addition, actually.

**Comment by vlsi** [comment]
>The repository in forRepository doesn't have to be created in the forRepository block:

Option 1:
```
repositories {
   maven { ... }
}

// if an exclusive filter is needed, the following must be typed:

repositories {
   val myRepo = maven { ... }
   exclusiveContent {
      forRepository { myRepo }
      filter { ... }
   }
}
```

Option 2:
```
repositories {
   maven { ... }
}

// vs

repositories {
   maven {
      ...
      exclusiveContent {
      }
   }
}
```

What I mean is it requires multiple actions: "declare a variable (invent name), add `exclusiveContent`, add `forRepository`, add `filter`", and only after that one can really add a filter.

If `exclusiveContent` block can be added right under repository declaration, then it would be more user-friendly.

**Comment by melix** [comment]
> If exclusiveContent block can be added right under repository declaration, then it would be more user-friendly.

Let me disagree on this. I think it's wrong to state at the repository level something which influences the resolution in other repositories. I understand _why_ you'd like that but I think it's not a good idea.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/RepositoryHandler.java [hunk 1]
```diff
@@ -324,4 +326,14 @@
      */
     IvyArtifactRepository ivy(Action<? super IvyArtifactRepository> action);
 
+    /**
+     * Defines the an exclusive content repository: whatever is declared in this
```
Comment (vlsi) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/RepositoryHandler.java]: the an

**Comment by vlsi** [comment]
> repository level something that says something about all repositories

You still use `exclusive content repository` wording in the javadoc for `exclusiveContent`  though: `Defines the an exclusive content repository`.

I see that you mean "it is better to avoid silent use of global variables", "it is better to avoid global consequences from configuring local entities".

However, the current PR does not enforce that `exclusiveContent` is declared the first, so it is not that different from `repository level something that says something about all repositories`

**Comment by melix** [comment]
Exclusive content doesn't have to be declared first. If you look at the tests I have actually tested the 2 orders. This is because it relies on `repositories.all` which doesn't care _when_ the repository is created.

I'd appreciate if you could come up with a test case which shows the order is relevant, if not too convoluted.

**Comment by vlsi** [comment]
> I'd appreciate if you could come up with a test case which shows the order is relevant

Oh, what I meant was even current design allows to declare "exclusiveContent" at the very end of the repository list, and it would influence the resolution of repositories that come before and/or after that.

In other words, even current implementation looks like having `at the repository level something that says something about all repositories`.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/DefaultRepositoryHandler.java [hunk 2]
```diff
@@ -146,4 +160,94 @@ public IvyArtifactRepository ivy(Action<? super IvyArtifactRepository> action) {
...
+            }
+            Set<? extends ArtifactRepository> targetRepositories = forRepositories.stream().map(Factory::create).collect(Collectors.toSet());
+            Action<? super RepositoryContentDescriptor> forExclusivity = transformForExclusivity(filter);
+            this.repositories.all(repo -> {
+                if (targetRepositories.contains(repo)) {
+                    repo.content(filter);
+                } else {
+                    repo.content(forExclusivity);
+                }
+            });
```
Comment (vlsi) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/DefaultRepositoryHandler.java]: Does this support the removal of the repository?

In other words, what if someone decides to remove a repository or remove `exclusiveContent` later?

Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/DefaultRepositoryHandler.java]: No it doesn't. It's a good point, although removing repositories is possible, it's not something recommended.

File: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc [hunk 3]
```diff
@@ -352,6 +353,26 @@ By default, repositories include everything and exclude nothing:
 It is possible to filter either by explicit _group_, _module_ or _version_, either strictly or using regular expressions.
 See link:{javadocPath}/org/gradle/api/artifacts/repositories/RepositoryContentDescriptor.html[RepositoryContentDescriptor] for details.
 
+=== Declaring content exclusively found in one repository
+
+Filters declared using the <<#sec:declaring-repository-filter,repository-level content filter>> are not exclusive.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc]: I would move this paragraph up, before the title. It concludes the section on content filters specific to a given repository.

Then, below the title, you have the rest that explains the purpose of exclusive filtering declaration.

Otherwise it feels weird to start a feature description by what another feature does not do.

**Comment by vlsi** [comment]
Here's an issue for IntelliJ IDEA to suggest users adding `exclusiveContent` for non-standard repositories: https://youtrack.jetbrains.com/issue/IDEA-299272
